### ISSUE TITLE
osrm-extract dropping ways (documentation request)

### ISSUE THREAD

**Issue by camilleanne** [Project-OSRM/osrm-backend#5574]
I'm working on running OSRM with a custom data set (prepared by ogr2osm). When I run `osrm-extract` I find that I have some surprising unconnected ways because they've been dropped from the graph. In the .osm file I've confirmed the existence of the dropped ways, and that they share nodes with ways that were preserved. At first I thought it was confusion around oneways, but I've found a few examples where that isn't a factor.

I've looked at this a million different ways and I don't see a reason why these ways didn't make it into the graph, and I can't find any difference between these intersections and intersections that are intact. Could someone explain the conditions for a way to be preserved/dropped? Or point me to a doc/relevant code?

<img width="587" alt="Screen Shot 2019-10-03 at 9 11 02 AM" src="https://user-images.githubusercontent.com/3952537/66142335-758b7280-e5c2-11e9-94eb-bb9ad147164e.png" >

*missing ways at intersection*

<img width="602" alt="Screen Shot 2019-10-03 at 9 54 16 AM" src="https://user-images.githubusercontent.com/3952537/66143163-ec753b00-e5c3-11e9-8c19-9858627fd85b.png">

*ways exist in .osm file*

<img width="707" alt="Screen Shot 2019-10-03 at 9 54 50 AM" src="https://user-images.githubusercontent.com/3952537/66143127-e4b59680-e5c3-11e9-9116-614630a57e72.png">
<img width="789" alt="Screen Shot 2019-10-03 at 9 54 33 AM" src="https://user-images.githubusercontent.com/3952537/66143153-ea12e100-e5c3-11e9-841d-bc9df6313ac7.png">
<img width="670" alt="Screen Shot 2019-10-03 at 9 54 29 AM" src="https://user-images.githubusercontent.com/3952537/66143155-eaab7780-e5c3-11e9-804e-db45a3597a8e.png">

nodes all seem in order:
```got a way with id  250726
nodes: [ -15252, -15253 ]
got a way with id  252264
nodes: [ -15253, -23652, -23653, -23654, -22886 ]
got a way with id  252639
nodes: [ -15253, -26343, -26344, -26345, -26346, -26347, -26348, -26349, -26350, -26351, -26352,  -26353, -11507 ]
```

what am I missing?

**Comment by danpat** [comment]
@camilleanne It needs to be `highway=trunk_link`, not `highway=trunk-link` - use an underscore, not a hyphen.

OSRM isn't recognizing the value, so it's dropping the ways completely.

**Comment by camilleanne** [comment]
@danpat 🙇‍♀ thanks very much. Out of curiosity, if I were to add `trunk-link` somewhere into the lua profile, would it keep those ways?

**Comment by danpat** [comment]
Yes - just search the profile for `trunk_link` and either change it, or add `trunk-link` to the lists - there are probably several places you need to do this, it's been a while since I looked.

IMO, you'd retain your sanity better by sticking to the published OSM standard tagging conventions :-)

**Comment by camilleanne** [comment]
Ok -- one follow up question then, it's related enough for this issue (more dropped ways because of tags):

When OSRM encounters a way without a `oneway` tag, it creates two parallel ways going in two directions. The dataset I'm using has ways with `toll=yes` and I'm finding that bidirectional ways with a `toll=yes` tag do not become two single direction ways each with `toll=yes` but instead becomes a single direction way (direction seems determined by node order). Ways with `oneway=yes` and `toll=yes` are unaffected.

If I knock out the `toll=yes` tags, everything looks fine.

Is it expected that the `toll=yes` tag knocks out bidirectional ways? I haven't been able to find anything that made me expect to treat toll roads differently except for making sure they are enabled in the lua profile.

<img width="971" alt="Screen Shot 2019-10-03 at 3 45 58 PM" src="https://user-images.githubusercontent.com/3952537/66167330-11cf6c80-e5f7-11e9-8936-14045ccddb53.png">

*no `toll=yes` tags* 

<img width="858" alt="Screen Shot 2019-10-03 at 3 52 29 PM" src="https://user-images.githubusercontent.com/3952537/66167336-13993000-e5f7-11e9-8a8b-10844872da5d.png">

*with `toll=yes` tags*

**Comment by danpat** [comment]
@camilleanne What happens when you try to calculate routes on these ways?  It's quite possible this is actually a bug in the debug visualization - the way the debug tiles are generated *might* be buggy when it comes to road classes that belong to the `excludes` sets.

Try changing this bit in `car.lua` to an empty sequence, regenerate your map data, and visualize again:

```
    -- classes to support for exclude flags
    excludable = Sequence {
        Set {'toll'},
        Set {'motorway'},
        Set {'ferry'}
    },
```

Change this to:

```
    -- classes to support for exclude flags
    excludable = Sequence {
        -- Set {'toll'},
        -- Set {'motorway'},
        -- Set {'ferry'}
    }
```

**Comment by camilleanne** [comment]
Doesn't seem like a bug in the debug tiles visualization since it's also represented in API response. Also, commenting out that bit doesn't change anything I can perceive -- I see this same behavior with or without it.

<img width="770" alt="Screen Shot 2019-10-03 at 3 57 18 PM" src="https://user-images.githubusercontent.com/3952537/66170884-4ea16080-e603-11e9-9d49-a5cc957d6847.png">

*routing in the correct direction*

<img width="901" alt="Screen Shot 2019-10-03 at 3 57 23 PM" src="https://user-images.githubusercontent.com/3952537/66170883-4ea16080-e603-11e9-871d-00297068a65e.png">

*reversed the trip -- trying to route all the way around the one way*

**Comment by danpat** [comment]
Hmm, this one has me stumped.

  1. Did you re-run `osrm-extract`, etc after changing the profile?  That needs to be done for profile changes.
  2. What other tags are present on the ways in question?
  3. Are your profiles customized from the defaults in any other ways?  What's in the `avoid=` setting in `car.lua` ?

There's nothing that I know of that `toll=yes` interacts with that should affect oneway/bidirectional behaviour.

**Comment by camilleanne** [comment]
1. Yes -- I reran everything (osrm-extract, osrm-partition, osrm-customize, osrm-routed). I double checked that the local lua was working by futzing with all the speeds and saw all the colors changed on the debug view.
2. other tags on two affected ways:

full_id | osm_id | osm_type | highway | id | lane | maxspeed | name | surface | toll | width | oneway | junction | layer | construction | service
 -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
w250222 | 250222 | way | motorway | 250222 | 4 | 100 | Cuacnopalan - Oaxaca | asphalt | yes | 12 |   |   |   |   |  
w249409 | 249409 | way | motorway | 249409 | 2 | 100 | Cuacnopalan - Oaxaca | asphalt | yes | 8 |   |   |   |   |  

3. I rolled back any changes I made to the lua profile and tried again commenting out the recommended lines, and brought them back. Both times the same.

**Comment by danpat** [comment]
`highway=motorway` implies that a road is oneway - see the panel on the right under "Implies" at https://wiki.openstreetmap.org/wiki/Tag:highway%3Dmotorway

The OSRM profile automatically does this here: https://github.com/Project-OSRM/osrm-backend/blob/master/profiles/lib/way_handlers.lua#L577-L588

It's the same logic that's applied to roundabouts - no need for an explicit oneway tag.

However, this doesn't quite explain the images in  https://github.com/Project-OSRM/osrm-backend/issues/5574#issuecomment-538147139 - they're showing that the road on the left, and the one heading north-east are both directional...sometimes.  Which roads correlate to the table in your previous comment?

**Comment by camilleanne** [comment]
fascinating -- I changed all the `highway=motorway` to `highway=trunk` and the `toll=yes` issue went away, bidirectional ways are bidirectional again with or without the toll tag.

I believe I pulled the `motorway` tag from the OSM Mexico wiki stating all toll roads should be motorways, and I didn't read the fine print.

However, yes it is weird that `highway=motorway` gets two edges without a `toll=yes` tag, and the addition of that tag triggers the implication of a one way. I just went and recreated it again to make sure I wasn't crazy.

here are the two sections referenced above (screenshot was taken after the highway tag was changed to `trunk` from `motorway`):
<img width="742" alt="Screen Shot 2019-10-03 at 7 04 41 PM" src="https://user-images.githubusercontent.com/3952537/66174174-fcffd280-e610-11e9-8280-e297735151e3.png">
<img width="755" alt="Screen Shot 2019-10-03 at 7 04 36 PM" src="https://user-images.githubusercontent.com/3952537/66174178-ff622c80-e610-11e9-88d6-16e3e1d44914.png">
